Insert images from local device (or link)

eneiluj opened the pull request:
Fixes #58

Here is a basic implementation of 2 new options to insert images:
* Upload from local device
* Download from http link

In both cases, the image is then stored in the `/Text` directory which is created if it does not exist.
An alternative mentioned in https://github.com/nextcloud/text/issues/58#issuecomment-540081015 would be to store uploaded images in a directory next to the markdown file.

### Todo
- [x] Menu action for local file insertion
- [x] Handling file upload
- [x] Display uploaded image
- [x] Handle and display upload errors
- [x] Menu action for image link insertion
- [x] Handling link download/store
- [x] Display image from link
- [x] Handle and display link insertion errors
- [x] Change menu item naming
- [x] Change attachment folder name
- [x] Use the attachment folder when inserting images from Files
- [x] Handle .md file copy/move/delete with File events
- [x] Cleanup attachment folder when images are removed from markdown content
### Gifcording :grin:
![link](https://user-images.githubusercontent.com/11291457/138112774-8dcb9f11-792d-40b0-9379-e20453acd43e.gif)

mejo- commented on src/components/MenuBar.vue:
```diff
@@ -23,6 +23,13 @@
+				accept="image/*"
+				aria-hidden="true"
+				class="hidden-visually"
+				@change="onImageFilePicked">
```
Nevermind, you already solved this :)

mejo- commented on lib/Controller/ImageController.php:
```diff
@@ -0,0 +1,86 @@
+		parent::__construct($appName, $request);
+		$this->userId = $userId;
+		$this->imageService = $imageService;
+		$this->request = $request;
```
Forget about that one, `$request` is incorporated from the parent class `Controller`.

eneiluj commented on src/components/MenuBar.vue:
```diff
@@ -264,33 +314,137 @@ export default {
+			// this avoids the input being reset on each file polling
+			this.imageLink = newImageLink
+		},
+		onImageLinksubmit(command) {
```
Yep! Thanks.

eneiluj commented:
This now includes "attachment folder" changes. It's been rebased on master (so it must be used in NC 24).

The TODO list in this PR's description has been updated.

eneiluj commented on src/components/MenuBar.vue:
```diff
@@ -34,6 +41,41 @@
 							:aria-label="t('text', 'Insert emoji')"
 							:aria-haspopup="true" />
 					</EmojiPicker>
+					<Actions v-else-if="icon.class === 'icon-image'"
```
I'll have a look.

eneiluj replied:
It's fine on my side with NC 24 :thinking: 
![col](https://user-images.githubusercontent.com/11291457/142861534-e52c0a2e-73a8-4914-bbec-a1f5df436bbb.png)

mejo- replied:
Interesting. I'm on latest NC 24 too :thinking: Will have to further look into it then.

mejo- replied:
I took another look and it seems like `vertical-align: top` is missing. The other action icons have it either through `menubar.button`, `menubar.submenu` or `menubar.menuitem-emoji` styling.

```suggestion
					<Actions v-else-if="icon.class === 'icon-image'"
						class="submenu"
```

eneiluj replied:
Oh I realized there actually is an alignment issue on my side too but it's the 3 dots menu icon rather than the image one. Anyway, using the submenu is fine. Done, pushed. Thanks!

juliushaertl commented on lib/Controller/ImageController.php:
```diff
@@ -74,6 +74,26 @@ public function __construct(string $appName,
+				return new DataResponse($insertResult);
+			}
+		} catch (Exception $e) {
+			return new DataResponse(['error' => 'File insertion error: ' . $e->getMessage()], Http::STATUS_BAD_REQUEST);
```
Let's not expose the actual exception message and rather log it or use a [HintException](https://github.com/nextcloud/server/blob/b78f3a57d1d05fb5f81ec8361139b64a3e54d7b0/lib/public/HintException.php#L36) for known cases. Otherwise exception messages might leak internal data.

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -394,6 +452,26 @@ private function getOrCreateAttachmentDirectoryForFile(File $textFile): ?Folder
+	private function getFileFromPath(string $filePath, string $userId): ?File {
+		$userFolder = $this->rootFolder->getUserFolder($userId);
+		if ($userFolder->nodeExists($filePath)) {
+			$file = $userFolder->get($filePath);
```
Will throw a NotFoundException so would be good to also annotate that if not catches

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,596 @@
+				return null;
+			}
+			if ($imageFile instanceof File) {
+//				return $imageFile;
```
- [ ] To be discussed if we still need the original file for gif support for example

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,596 @@
+				return null;
+			}
+			if ($imageFile instanceof File) {
+//				return $imageFile;
```
- [ ] To be discussed if we still need the original file for gif support for example

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,596 @@
+			$fileName = (string) time() . '-' . $imageFile->getName();
+			$targetPath = $saveDir->getPath() . '/' . $fileName;
+			$targetFile = $imageFile->copy($targetPath);
+			$path = preg_replace('/^files/', '', $targetFile->getInternalPath());
```
Might not work well with other storages than the user home (e.g. group folders)

I think something like this should also work to resolve the file within the user folder:

`$userFolder->getRelativePath($file->getPath())`

eneiluj replied:
Oh right. I'll keep that in mind. In this specific case, the `path` attribute is actually not used anymore in the frontend. It has been removed.

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,596 @@
+				'id' => $targetFile->getId(),
+				'textFileId' => $textFile->getId(),
+			];
+		} else {
+			return [
+				'error' => 'Unsupported file type',
+			];
+		}
```
```suggestion
	return [
		'error' => 'Unsupported file type',
	];
```

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,596 @@
+	private function downloadLink(Folder $saveDir, string $link, File $textFile): array {
+		$fileName = (string) time();
+		$savedFile = $saveDir->newFile($fileName);
+		$resource = $savedFile->fopen('w');
```
```suggestion
		$resource = $savedFile->fopen('wb');
```

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,596 @@
+				}
+			}
+
+			if ($method === 'GET') {
+				$response = $client->get($url, $options);
+			} elseif ($method === 'POST') {
+				$response = $client->post($url, $options);
+			} elseif ($method === 'PUT') {
+				$response = $client->put($url, $options);
+			} elseif ($method === 'DELETE') {
+				$response = $client->delete($url, $options);
+			} else {
+				return ['error' => 'Bad HTTP method'];
+			}
```
There doesn't seem to be much use of other methods than GET here, right?

eneiluj replied:
Yes and we most likely won't use other methods later.

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,596 @@
+			$this->logger->error('Connection error: ' . $e->getMessage(), ['app' => Application::APP_NAME]);
+			return ['error' => 'Connection error'];
+		} catch (Throwable | Exception $e) {
+			$this->logger->error('Unknown error: ' . $e->getMessage(), ['app' => Application::APP_NAME]);
```
```suggestion
			$this->logger->error('Unknown error when downloading the file', ['exception' => $e]);
```

No need for specifying the app if the LoggerInterface is injected through DI as it will be resolved to an instance that has the app already set. Also losing the exception directly will preserve the trace which might be helpful.

eneiluj replied:
Ok, fixed in many places.

juliushaertl commented on src/components/MenuBar.vue:
```diff
@@ -225,6 +294,10 @@ export default {
 			return this.lastImagePath
 				|| this.filePath.split('/').slice(0, -1).join('/')
 		},
+		token() {
```
Might be good to call it sharingToken, as there is a separate token for the session within the text codebase as well, to avoid any confusion

juliushaertl commented on lib/Controller/ImageController.php:
```diff
@@ -0,0 +1,246 @@
+	 * @param string $imagePath
+	 * @return DataResponse
+	 */
+	public function insertImageFile(int $textFileId, string $imagePath): DataResponse {
```
Thinking about that it might be useful to just make this a public endpoint for both logged in users and public share links and use the session token to authenticate. That way the same methods can also be used when opening a file through direct editing (ref https://github.com/nextcloud/server/pull/17625) through the desktop/Android/iOS clients.

FYI @tobiasKaminsky For inserting local files/files from an URL this should then work without any client change. The existing feature to insert an image from Nextcloud would though depend on https://github.com/nextcloud/text/issues/1170

eneiluj replied:
@juliushaertl Yes sure, seems better. Do you have any example where authentication is checked like that?

mejo- commented on lib/Service/ApiService.php:
```diff
@@ -47,12 +47,18 @@ class ApiService {
 	protected $sessionService;
 	protected $documentService;
 	protected $logger;
+	private ImageService $imageService;
```
That breaks on PHP7.3 :)

```suggestion
	protected $imageService;
```

eneiluj replied:
Right, I though Php 7.3 support had been dropped in NC 24 but it's actually not.

eneiluj commented:
Hey there, some news:
* Attachment **cleanup** is implemented. It happens when closing Text if there are no more open sessions. It's also done by the Cleanup cron job when cleaning up the inactive sessions.
* The attachment dir is **deleted** when a markdown file is deleted.
* The attachment dir is **moved** to follow the markdown file when it's moved. This is only done if the moved .md file was in the same directory than the attachment dir. We make sure the attachment dir stays next to the .md file in the owner's storage. This covers all the use case I could think of:
    * the .md file's owner moves it: the attachment dir is moved
    * the .md file is shared (single file share) and moved by the share target user: the attachment is **not** moved
    * a parent directory of the .md file is shared, the share target user moves the .md file: the attachment dir is moved.
* File **copy** is implemented but there is a major issue: The target file is locked when catching the event so we can't make changes to its content (which is needed to update the attachment links). @juliushaertl @mejo- Any idea how we could proceed here?

eneiluj commented:
Some news here:
* Attachment cleanup is done:
    * when closing the document and there is no other active session
    * in the Cleanup cron job if an inactive session is being deleted and there is no other active ones
 * On file copy, the attachment folder is copied too. 
 * On file move, the attachment folder is moved only if the .md file was in its owner's storage.
 * On file deletion, the attachment folder is deleted.

So I think everything we discussed is implemented. There still are some things to check and improve:

- [ ] We need to edit the target .md file when copying it to adapt the attachment file IDs in its content. Problem is the .md file is locked when we catch the copy event. For now it's temporarily unlocked but I'm sure there is a cleaner solution :grin:
- [ ] The .md file ID is stored in each attachment link in order to use it as parameter when getting images through Text. This is kind of hacky. It would be better to find a way to get the .md file ID in the ImageView component context.
- [ ] As mentioned earlier, it would be better to have a single public endpoint to serve images to logged in users and public share links. The session token would be used to authenticate.

eneiluj commented:
Rebased on master.

* @juliushaertl I tried to implement authentication via edition session like you said. There is now only one public endpoint for each upload method. In non-public contexts, we now get the userId from the session instead of relying on NC auth.
Could you have a look at https://github.com/nextcloud/text/pull/1900/commits/06d4d3866caf7c9c9de65b31ee09e22b9721b934 ? Is this more or less what you had in mind?

* It would be nice to use the edition session when serving images too. It would allow us to avoid storing the document ID in each image link (which is...meh). But I can't find a clean way to get the session info in the ImageView component to be able to put it in the "get image" requests.

eneiluj commented:
I think it's finally in a good shape :tada: .

Thanks to @mejo- 's suggestion, images are now served via a single endpoint in both public and non-public contexts. Session information is now accessible via the Vuex store so the ImageView component can use it to generate the image URL.

@juliushaertl @mejo- I wonder if those changes to the Vuex store are ok: 7e0df82d076a6dcaf1c20babcbfb45ef542e9277 more specifically in `src/nodes/ImageView.vue` and `src/store.js`.

The document file ID and the image file IDs are not written in the markdown content anymore which makes it easier to handle the .md file copy/move/delete events. Only the attachment image file name is written.

Oh and I don't think the failing CI checks are blocking :grin:.

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+			if ($imageFile instanceof File) {
+				if ($this->previewManager->isMimeSupported($imageFile->getMimeType())) {
+					return $this->previewManager->getPreview($imageFile, 1024, 1024);
+				} else {
```
From PhpStorm: [EA] Child instructions can be extracted here (clearer intention, lower complexity numbers).

eneiluj replied:
I don't get this on my side. What does it mean? :grin:

mejo- replied:
It basically means that you can remove the `else {` statement and instead put everything inside `else` **behind** the `if` statement.

It's about changing

```php
if {
    ...
    return x;
} else {
    do_something();
}
```

to

```php
if {
    ...
    return x;
}
do_something();
```

mejo- replied:
See https://github.com/kalessil/phpinspectionsea/blob/master/docs/control-flow.md#redundant-else-keyword for details :blush:

I guess it's the setting in Editor -> Inspections -> PHP -> Control flow -> Unnecessary 'else' branch.

eneiluj replied:
:+1: thanks

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+	public function uploadImage(int $documentId, string $newFileName, string $newFileContent, string $userId): array {
+		$textFile = $this->getTextFile($documentId, $userId);
+		if (!$textFile->isUpdateable()) {
+			throw new Exception('No write permissions');
```
Maybe throw a `\OCP\Files\NotPermittedException` here instead of a generic `\Exception`?

eneiluj replied:
:+1:

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+	 */
+	public function uploadImage(int $documentId, string $newFileName, string $newFileContent, string $userId): array {
+		$textFile = $this->getTextFile($documentId, $userId);
+		if (!$textFile->isUpdateable()) {
```
Better to check for $textFile to prevent null pointer exception:

```suggestion
		if (!$textFile || !$textFile->isUpdateable()) {
```

Or even better, throw a `NotFoundException` if empty in `$this->getTextFile()`?

eneiluj replied:
Yep. `ImageController::getImage()` now catches the exceptions.
`ImageService::getTextFile()` and `ImageService::getTextFilePublic()` now throw a `NotFoundException`.

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+				'id' => $savedFile->getId(),
+				'documentId' => $textFile->getId(),
+			];
+		} else {
```
From PhpStorm: [EA] Child instructions can be extracted here (clearer intention, lower complexity numbers).

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	private function getOrCreateTextDirectory(string $userId): ?Folder {
```
This seems like an old leftover? At lest the function is not used.

eneiluj replied:
:+1:

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+					'id' => $savedFile->getId(),
+					'documentId' => $textFile->getId(),
+				];
+			} else {
```
From PhpStorm: [EA] Child instructions can be extracted here (clearer intention, lower complexity numbers).

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+		$saveDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($saveDir !== null) {
+			return $this->downloadLink($saveDir, $link, $textFile);
+		} else {
```
From PhpStorm: [EA] Child instructions can be extracted here (clearer intention, lower complexity numbers).

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+		$saveDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($saveDir !== null) {
+			return $this->downloadLink($saveDir, $link, $textFile);
+		} else {
```
From PhpStorm: [EA] Child instructions can be extracted here (clearer intention, lower complexity numbers).

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+	 */
+	public function insertImageLink(int $documentId, string $link, string $userId): array {
+		$textFile = $this->getTextFile($documentId, $userId);
+		if (!$textFile->isUpdateable()) {
```
Same as above: either check to prevent null pointer exception, or throw in `$this->getTextFile()`.

mejo- reviewed (changes requested):
In general, I'm really happy with the implemention and think it's almost done. All my comments are rather minor compared to the huge PR.

But how about adding some cypress end-to-end tests for image uploads?

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+	private function getAttachmentNamesFromContent(string $content): array {
+		$matches = [];
+		preg_match_all(
+			'/\!\[[^\[\]]+\]\(text:\/\/image\?[^)]*imageFileName=([^)&]+)\)/',
```
The regex doesn't match image links with square brackets in the description. I had a quick look at [the relevant part of the CommonMark spec](https://spec.commonmark.org/0.30/#links) and indeed, square brackets are allowed (under limited circumstances).

php-markdown uses the following regex syntax for the link/image name/description part between squares:

```regex
(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[\])*\])*\])*\])*\])*\])*
```

See [`$this->nested_brackets_re`](https://github.com/michelf/php-markdown/blob/f0911e737ddaad081803424203407a0f01189eca/Michelf/Markdown.php#L106) and [`doImages()`](https://github.com/michelf/php-markdown/blob/f0911e737ddaad081803424203407a0f01189eca/Michelf/Markdown.php#L693)

eneiluj replied:
This is tricky. The inserted `[photo.name.jpg](text://image?imageFileName=photo.name.jpg)` has to be valid from the renderer point of view. I'll first fix that and then adjust `getAttachmentNamesFromContent()`. Thanks for mentioning this as image insertion fails in many cases.

eneiluj replied:
The expression from php-markdown is now used in `getAttachmentNamesFromContent()`.
As a simple fix for the nested brackets, as we generate the link ourselves and the user is not supposed to manually edit it afterwards, brackets in image file names are removed in the link text.
There is no problem in the link itself as it is "url encoded".
This is done in c351e66607d23b1672603752b2bdd1dba83b59a6

mejo- replied:
Perfect, I think that's fine for now :ok_hand:

eneiluj commented on src/components/MenuBar.vue:
```diff
@@ -34,6 +41,43 @@
+							:aria-haspopup="true" />
+						<ActionButton
+							icon="icon-upload"
+							:close-after-click="true"
```
:+1:

eneiluj commented on src/components/MenuBar.vue:
```diff
@@ -34,6 +41,43 @@
+							@click="showImagePrompt(commands.image)">
+							{{ t('text', 'Insert from Files') }}
+						</ActionButton>
+						<ActionButton v-show="!showImageLinkPrompt"
```
Yes, right. I remember having some issues with v-if in <Action*> where an action is not visible if it is rendered after <Actions>. But here it seems to work fine.

eneiluj commented on lib/AppInfo/Application.php:
```diff
@@ -53,6 +59,10 @@ public function register(IRegistrationContext $context): void {
+		// for image attachments
+		$context->registerEventListener(NodeCopiedEvent::class, NodeCopiedListener::class);
+		$context->registerEventListener(BeforeNodeRenamedEvent::class, BeforeNodeRenamedListener::class);
+		$context->registerEventListener(BeforeNodeDeletedEvent::class, BeforeNodeDeletedListener::class);
```
It's not possible to get the attachment dir after a file has been deleted or moved because we use it (the file) to get its owner's storage.
I agree it would be nicer to use NodeRenamedEvent and BeforeNodeDeletedEvent but I think it's not possible.

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	private function getOrCreateAttachmentDirectoryForFile(File $textFile): ?Folder {
```
Since `getOrCreateAttachmentDirectoryForFile()` and `getAttachmentDirectoryForFile()` share a lot of code, wouldn't it be better to merge them and instead e.g. add a boolean parameter `$create`?

eneiluj replied:
:+1:

eneiluj commented:
@mejo- Thanks a lot for the review.

eneiluj commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+			$matches,
+			PREG_SET_ORDER
+		);
+		return array_map(function (array $match) {
```
:+1:

eneiluj commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+				// get IDs of the files inside the attachment dir
+				$attachmentDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+				$attachmentsByName = [];
+				foreach ($attachmentDir->getDirectoryListing() as $attNode) {
```
Yep. Moreover, we don't need to create the attachment dir there if it does not exist.

eneiluj commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+			// if there is an attachment dir for this file
+			// and it is in the same directory as the source file
+			if ($sourceAttachmentDir !== null
+				&& $source->getParent()->getId() === $sourceAttachmentDir->getParent()->getId()
```
I improved the comment. This is tricky.
* move:
    * if done by the .md file's owner, the attachment dir is next to the .md file => we should move the attachment dir
    * if someone else having access does it, as the file won't move in the owner's storage => we don't move the attachment dir
* copy is different because we duplicate the attachments, this is done regardless if the file is copied by its owner or someone else.

Does it make sense? :grin:

mejo- replied:
Ah, makes sense, thanks for the explanation! Maybe just use the last comment?

```php
		// Move attachement dir only if .md file is owned by it's owner.
```

eneiluj commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+		}
+	}
+
+	public function copyAttachments(File $source, File $target): void {
```
:+1:

eneiluj commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,676 @@
+					if ($textFile instanceof File) {
+						return $textFile;
+					}
+				} elseif ($share->getNodeType() === 'folder' && $documentId !== null) {
```
:+1:

eneiluj commented on lib/Controller/ImageController.php:
```diff
@@ -0,0 +1,224 @@
+	 * @throws \OCP\Files\NotPermittedException
+	 * @throws \OCP\Lock\LockedException
+	 */
+	public function getImage(int $documentId, int $sessionId, string $sessionToken, ?string $shareToken = null, string $imageFileName): DataDisplayResponse {
```
:+1:

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,658 @@
+				return $ownerParentFolder->newFolder($attachmentFolderName);
+			}
+		}
+		return null;
```
I wonder whether it would be cleaner to throw a `NotFoundException` instead of returning null in low-level functions like this one, `getFileFromPath()` and `getImagePreview()`. That way you don't have to check for the returned value being null in the higher level functions and don't have to care about null pointer exceptions.

eneiluj replied:
Much better, simpler and more readable.
Exceptions are now caught in the controller (except for the delete/copy/move methods in the service).
Done in c204d904ab3a1bba6c58d32d8a615ebab04808b4

mejo- commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,658 @@
+	 * @param array $params
+	 * @return array
+	 */
+	private function simpleDownload(string $url, $resource, array $params = []): array {
```
The parameter `$params` is never passed to the private function `simpleDownload()` and probably can be dropped, no?

eneiluj replied:
:+1:

eneiluj commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,658 @@
+	 */
+	private function copyImageFile(File $imageFile, Folder $saveDir, File $textFile): array {
+		$mimeType = $imageFile->getMimeType();
+		if (in_array($mimeType, ImageController::IMAGE_MIME_TYPES)) {
```
:+1:

eneiluj commented on lib/Controller/ImageController.php:
```diff
@@ -0,0 +1,225 @@
+		try {
+			$file = $this->request->getUploadedFile('image');
+			if ($file !== null && isset($file['tmp_name'], $file['name'], $file['type'])) {
+				if (!in_array($file['type'], self::IMAGE_MIME_TYPES)) {
```
:+1:

eneiluj commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,658 @@
+		if (isset($res['Content-Type'])) {
+			if (in_array($res['Content-Type'], ImageController::IMAGE_MIME_TYPES)) {
+				if ($res['Content-Type'] === 'image/jpeg') {
+					$fileName = $fileName . '.jpg';
```
:+1:

juliushaertl commented on lib/Controller/ImageController.php:
```diff
@@ -0,0 +1,210 @@
+
+		try {
+			$file = $this->request->getUploadedFile('image');
+			if ($file !== null && isset($file['tmp_name'], $file['name'], $file['type'])) {
```
Some validation on the uploaded file would be nice here, to give some more feedback to the user or log, for possible inspiration: https://github.com/nextcloud/deck/blob/13dcacc3bb4149ebc7283f34f6d9af05322adea5/lib/Service/FilesAppService.php#L218-L235 but not a blocker for me.

eneiluj replied:
:+1: Done in bef8698e291caab8ff5e0485e9b2b53f63aa888c (almost like in Deck).

juliushaertl commented on lib/Controller/ImageController.php:
```diff
@@ -0,0 +1,210 @@
+				if (!in_array($file['type'], self::IMAGE_MIME_TYPES, true)) {
+					return new DataResponse(['error' => 'Image type not supported'], Http::STATUS_BAD_REQUEST);
+				}
+				$newFileContent = file_get_contents($file['tmp_name']);
```
Also no blocker but would be good to stream the file to the storage instead of loading it into memory. https://github.com/nextcloud/deck/blob/13dcacc3bb4149ebc7283f34f6d9af05322adea5/lib/Service/FilesAppService.php#L194-L198

eneiluj replied:
:+1: Done in d1cf1fb7f80b68bf3f3af0769c5e410ba831e2c2

juliushaertl commented on lib/Controller/ImageController.php:
```diff
@@ -0,0 +1,210 @@
+				$imageFile = $this->imageService->getImage($documentId, $imageFileName, $this->userId);
+			}
+			return $imageFile !== null
+				? new DataDisplayResponse($imageFile->getContent(), Http::STATUS_OK, ['Content-Type' => $imageFile->getMimeType()])
```
I'm not fuilly sure if necessary, but to be on the safe side this should use https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/public/Files/IMimeTypeDetector.php#L68 to avoid any mimetype that would allow potential XSS

eneiluj replied:
:+1: Done in 7628714403f41706f5dd9465c7a4b48898a6ecbc

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,631 @@
+	 * @param string $content
+	 * @return array
+	 */
+	private function getAttachmentNamesFromContent(string $content): array {
```
Bonus points for some quick unit test ;)

eneiluj replied:
Yep, will do after fixing smaller stuff.

eneiluj replied:
A small test was added in c351e66607d23b1672603752b2bdd1dba83b59a6
Do you have more ideas of potentially problematic names?

juliushaertl replied:
Looks good :)

juliushaertl commented on lib/Service/ImageService.php:
```diff
@@ -0,0 +1,631 @@
+	private function hasUpdatePermissions(string $shareToken): bool {
+		try {
+			$share = $this->shareManager->getShareByToken($shareToken);
+			return ($share->getShareType() === IShare::TYPE_LINK && $share->getPermissions() & Constants::PERMISSION_UPDATE);
```
`IShare::TYPE_EMAIL` would also be a case where this should work as they behave the same like share links.

juliushaertl replied:
`IShare::TYPE_ROOM` would also be possible if the room is shared to the public, then such shares would be used to access the file as guests

eneiluj replied:
:+1: Done in fad15fa0a42010eab2905dba2e9479d1a53ca025

juliushaertl reviewed (changes requested):
Very nice work @eneiluj I've added a few comments inline, but that should be rather small stuff.

juliushaertl commented:
Regarding the psalm issues on CI, the following two can be added to the baseline as they are due to leaking oc namespace classes in ocp ones: `oc\hooks\emitter` and `OC\User\NoUserException`

eneiluj commented on lib/Controller/ImageController.php:
```diff
@@ -0,0 +1,210 @@
+	 */
+	public function insertImageLink(string $link, int $documentId, int $sessionId, string $sessionToken, ?string $shareToken = null): DataResponse {
+		if (!$this->sessionService->isValidSession($documentId, $sessionId, $sessionToken)) {
+			return new DataResponse([], Http::STATUS_INTERNAL_SERVER_ERROR);
```
:+1:

mejo- commented on lib/Controller/ImageController.php:
```diff
@@ -175,6 +183,35 @@ public function uploadImage(int $documentId, int $sessionId, string $sessionToke
+			$error = $phpFileUploadErrors[$file['error']];
+		}
+		if ($error !== null) {
+			throw new Exception($error);
```
Maybe again a more specific exception would be nice here? :thinking:

eneiluj replied:
Yep, done in c92316183af72ee92a6a73a2b31e4b7da0255c74

eneiluj commented:
@juliushaertl @mejo- I think everything you mentioned is fixed.

I'll rebase, squash the commits and compile when we're done with the review.

mejo- commented:
Super nice work :ok_hand:

As mentioned, some cypress tests would be nice - apart from that, the PR is ready to be merged from my side :tada:

eneiluj commented:
> Super nice work ok_hand
> 
> As mentioned, some cypress tests would be nice - apart from that, the PR is ready to be merged from my side tada

Oh yes right, I'll dive into cypress soon.
Thanks again for the solid review!

eneiluj commented:
Rebased on master, added small tests for `<ImageView>`.
We can now merge or wait until a pairing session about Cypress and some more tests. As you wish.

juliushaertl commented:
Good to merge right away from my side 👍

eneiluj commented:
/compile amend

mejo- commented:
> We can now merge or wait until a pairing session about Cypress and some more tests. As you wish.

I don't mind to merge now and add cypress tests later (if we don't forget about it :grimacing:). But the failing psalm tests should be tackled first, no?

eneiluj commented:
/compile amend

mejo- reviewed (approved):
Super nice work :ok_hand:

juliushaertl commented:
Filed https://github.com/nextcloud/text/issues/2049 for the cypress tests

tobiasKaminsky commented:
Is this also ready for clients via our DirectEditingMobileInterface?

eneiluj commented:
@tobiasKaminsky All related endpoints are public and use the session edition token and optionally the public share token to authenticate.
So it is supposed to be ok but in reality it partially works. I've tried it on our Android app:
* Inserting an image from a link :heavy_check_mark: 
* Uploading a local image :x: The `<input type=file>` does not work when rendered in the Android app.
* Serving the images :x: 2 mistakes:
    * The share token presence detection is broken in <ImageView>
    * The ImageController does not get the user ID from the session

I'll make a PR to fix the 2 image serving issues.
I don't know about the local image upload issue. Do you think we should just hide this feature on mobile apps? Or is there a way to get a system file selector when clicking on such file input?

tobiasKaminsky commented:
> I don't know about the local image upload issue. Do you think we should just hide this feature on mobile apps? Or is there a way to get a system file selector when clicking on such file input?

This is possible, as we do it with Collabora.

In this regard, we have this https://github.com/nextcloud/server/issues/16132, but want to wait for your side (move RichDocuments to directEditing)
